sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -359,20 +386,129 @@
+        }
+      return ImmutableList.copyOf(topicPartitionList);
+      }
+    }
+
+    /**
+     * Sets a Kafka {@link Deserializer Deserializer&lt;K&gt;} for interpreting key bytes read.
+     * This uses the {@link String} provided to set the Deserializer
+     */
+    public Read<K, V> withKeyDeserializer(String keyDeserializer) {
+      return withKeyDeserializer(ValueProvider.StaticValueProvider.of(keyDeserializer));
+    }
+
+    /**
+     * Like above but with a {@link ValueProvider ValueProvider&lt;String&gt;}.
+     */
+    public Read<K, V> withKeyDeserializer(ValueProvider<String> keyDeserializer) {
+      return toBuilder().setKeyDeserializer(ValueProvider
+              .NestedValueProvider.of(keyDeserializer, new KeyDeserializerTranslator())).build();
     }
 
     /**
-     * Sets a Kafka {@link Deserializer} to interpret key bytes read from Kafka.
+     * Sets a Kafka {@link Deserializer Deserializer&lt;K&gt;} to interpret key bytes read.
      *
      * <p>In addition, Beam also needs a {@link Coder} to serialize and deserialize key objects at
      * runtime. KafkaIO tries to infer a coder for the key based on the {@link Deserializer} class,
      * however in case that fails, you can use {@link #withKeyDeserializerAndCoder(Class, Coder)} to
      * provide the key coder explicitly.
      */
     public Read<K, V> withKeyDeserializer(Class<? extends Deserializer<K>> keyDeserializer) {
-      return toBuilder().setKeyDeserializer(keyDeserializer).build();
+      return toBuilder().setKeyDeserializer(
+          ValueProvider.StaticValueProvider.of(keyDeserializer)).build();
+    }
+
+    /**
+     * Used to build a {@link ValueProvider} for {@link Deserializer Deserializer&lt;K&gt;}.
+     */
+    private class KeyDeserializerTranslator
+        implements SerializableFunction<String, Class<? extends Deserializer<K>>> {
+      @SuppressWarnings("unchecked")
+      @Override
+      public Class apply(String deserializer) {
+        Class deserializerClass;
+        try {
+          deserializerClass =  Class.forName(deserializer);
+        } catch (ClassNotFoundException e) {
+          LOG.error("Provided class for KeySerializer not found {}", deserializer);
+          throw new RuntimeException("Please check for the existence of the KeySerializer Class",
+              e);
+        }
+        return deserializerClass;
+      }
+    }
+
+    /**
+     * Used to build a {@link ValueProvider} for {@link Deserializer Deserializer&lt;V&gt;}.
+     */
+    private class ValueDeserializerTranslator
+            implements SerializableFunction<String, Class<? extends Deserializer<V>>> {
+      @SuppressWarnings("unchecked")
+      @Override
+      public Class apply(String deserializer) {
+        Class deserializerClass;
+        try {
+          deserializerClass =  Class.forName(deserializer);
+        } catch (ClassNotFoundException e) {
+          LOG.error("Provided class for KeySerializer not found {}", deserializer);
```

Comment:
Ditto. Besides:
- copy-paste typo: value deserializer
- KeyDeserializerTranslator and ValueDeserializerTranslator are identical, you only need one (but named appropriately, e.g. package it as a `ValueProviders.classForName` utility function)

Reply:
I have added classForName method inside. ValueProviders. 

How can I merge the two below when one extends Deserializer<K> and another extends Deserializer<V>? Please provide guidance.

private class KeyDeserializerTranslator
        implements SerializableFunction<String, Class<? extends Deserializer<K>>> and 

private class ValueDeserializerTranslator
        implements SerializableFunction<String, Class<? extends Deserializer<V>>>